File: backend/auth_service.py
import os
from datetime import datetime, timedelta, timezone
from typing import Optional
import uuid

from passlib.context import CryptContext
from jose import JWTError, jwt

from database import users_col, refresh_tokens_col, provider_wallets_col, patient_profiles_col

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

ACCESS_SECRET = os.environ.get("JWT_ACCESS_SECRET")
REFRESH_SECRET = os.environ.get("JWT_REFRESH_SECRET")
ACCESS_EXPIRY = int(os.environ.get("JWT_ACCESS_EXPIRY_MINUTES", "15"))
REFRESH_EXPIRY = int(os.environ.get("JWT_REFRESH_EXPIRY_DAYS", "7"))


def hash_password(password: str) -> str:
    return pwd_context.hash(password)


def verify_password(plain: str, hashed: str) -> bool:
    return pwd_context.verify(plain, hashed)


def create_access_token(user_id: str, role: str) -> str:
    payload = {
        "sub": user_id,
        "role": role,
        "type": "access",
        "exp": datetime.now(timezone.utc) + timedelta(minutes=ACCESS_EXPIRY),
        "iat": datetime.now(timezone.utc),
    }
    return jwt.encode(payload, ACCESS_SECRET, algorithm="HS256")


def create_refresh_token(user_id: str) -> str:
    payload = {
        "sub": user_id,
        "type": "refresh",
        "jti": str(uuid.uuid4()),
        "exp": datetime.now(timezone.utc) + timedelta(days=REFRESH_EXPIRY),
        "iat": datetime.now(timezone.utc),
    }
    return jwt.encode(payload, REFRESH_SECRET, algorithm="HS256")


def decode_access_token(token: str) -> Optional[dict]:
    try:
        payload = jwt.decode(token, ACCESS_SECRET, algorithms=["HS256"])
        if payload.get("type") != "access":
            return None
        return payload
    except JWTError:
        return None


def decode_refresh_token(token: str) -> Optional[dict]:
    try:
        payload = jwt.decode(token, REFRESH_SECRET, algorithms=["HS256"])
        if payload.get("type") != "refresh":
            return None
        return payload
    except JWTError:
        return None


async def register_user(email: str, password: str, first_name: str, last_name: str, role: str, phone: str = None) -> dict:
    existing = await users_col.find_one({"email": email})
    if existing:
        raise ValueError("Email already registered")

    user_id = str(uuid.uuid4())
    now = datetime.now(timezone.utc).isoformat()

    user_doc = {
        "id": user_id,
        "email": email,
        "password_hash": hash_password(password),
        "first_name": first_name,
        "last_name": last_name,
        "role": role,
        "phone": phone,
        "is_active": True,
        "email_verified": False,
        "created_at": now,
        "updated_at": now,
    }
    await users_col.insert_one(user_doc)

    # Create role-specific profiles
    if role in ("THERAPIST", "PSYCHIATRIST", "COACH"):
        wallet_doc = {
            "id": str(uuid.uuid4()),
            "provider_id": user_id,
            "available_balance": 0,
            "pending_balance": 0,
            "lifetime_earnings": 0,
            "total_withdrawn": 0,
            "created_at": now,
            "updated_at": now,
        }
        await provider_wallets_col.insert_one(wallet_doc)

    if role == "PATIENT":
        patient_doc = {
            "id": str(uuid.uuid4()),
            "user_id": user_id,
            "subscription_tier": "FREE",
            "primary_concerns": [],
            "preferred_languages": ["en"],
            "consent_given": False,
            "created_at": now,
            "updated_at": now,
        }
        await patient_profiles_col.insert_one(patient_doc)

    return _safe_user(user_doc)


async def login_user(email: str, password: str) -> dict:
    user = await users_col.find_one({"email": email})
    if not user or not verify_password(password, user["password_hash"]):
        raise ValueError("Invalid email or password")

    if not user.get("is_active", True):
        raise ValueError("Account is deactivated")

    # Update last login
    await users_col.update_one(
        {"id": user["id"]},
        {"$set": {"last_login_at": datetime.now(timezone.utc).isoformat()}}
    )

    access = create_access_token(user["id"], user["role"])
    refresh = create_refresh_token(user["id"])

    # Store refresh token
    await refresh_tokens_col.insert_one({
        "id": str(uuid.uuid4()),
        "user_id": user["id"],
        "token": refresh,
        "expires_at": datetime.now(timezone.utc) + timedelta(days=REFRESH_EXPIRY),
        "created_at": datetime.now(timezone.utc).isoformat(),
    })

    return {
        "access_token": access,
        "refresh_token": refresh,
        "token_type": "bearer",
        "user": _safe_user(user),
    }


async def refresh_tokens(refresh_token: str) -> dict:
    payload = decode_refresh_token(refresh_token)
    if not payload:
        raise ValueError("Invalid refresh token")

    stored = await refresh_tokens_col.find_one({"token": refresh_token})
    if not stored or stored.get("revoked_at"):
        raise ValueError("Refresh token revoked or not found")

    # Revoke old token (rotation)
    await refresh_tokens_col.update_one(
        {"token": refresh_token},
        {"$set": {"revoked_at": datetime.now(timezone.utc).isoformat()}}
    )

    user = await users_col.find_one({"id": payload["sub"]})
    if not user:
        raise ValueError("User not found")

    access = create_access_token(user["id"], user["role"])
    new_refresh = create_refresh_token(user["id"])

    await refresh_tokens_col.insert_one({
        "id": str(uuid.uuid4()),
        "user_id": user["id"],
        "token": new_refresh,
        "expires_at": datetime.now(timezone.utc) + timedelta(days=REFRESH_EXPIRY),
        "created_at": datetime.now(timezone.utc).isoformat(),
    })

    return {
        "access_token": access,
        "refresh_token": new_refresh,
        "token_type": "bearer",
        "user": _safe_user(user),
    }


async def logout_user(refresh_token: str):
    await refresh_tokens_col.update_one(
        {"token": refresh_token},
        {"$set": {"revoked_at": datetime.now(timezone.utc).isoformat()}}
    )


def _safe_user(user: dict) -> dict:
    return {
        "id": user["id"],
        "email": user["email"],
        "first_name": user["first_name"],
        "last_name": user["last_name"],
        "role": user["role"],
        "phone": user.get("phone"),
        "is_active": user.get("is_active", True),
        "created_at": user.get("created_at"),
    }

File: backend/src/modules/auth/auth.service.ts
import { PrismaClient, UserRole } from '@prisma/client';
import jwt from 'jsonwebtoken';
import { cryptoUtils } from '../../utils/crypto';
import { env } from '../../config/env';
import { AppError } from '../../middleware/error.middleware';
import { ERROR_CODES, HTTP_STATUS } from '../../utils/constants';
import { logger } from '../../utils/logger';

const prisma = new PrismaClient();

interface RegisterInput {
  email: string;
  password: string;
  firstName: string;
  lastName: string;
  role: UserRole;
  phone?: string;
}

interface LoginInput {
  email: string;
  password: string;
  ipAddress?: string;
  userAgent?: string;
}

interface RefreshTokenInput {
  refreshToken: string;
  ipAddress?: string;
  userAgent?: string;
}

interface TokenPayload {
  userId: string;
  email: string;
  role: UserRole;
}

export class AuthService {
  /**
   * Register new user
   */
  async register(input: RegisterInput) {
    // Check if user exists
    const existing = await prisma.user.findUnique({
      where: { email: input.email },
    });

    if (existing) {
      throw new AppError('User already exists', HTTP_STATUS.CONFLICT, ERROR_CODES.VALIDATION_ERROR);
    }

    // Hash password
    const passwordHash = await cryptoUtils.hashPassword(input.password);

    // Create user and profile in transaction
    const user = await prisma.$transaction(async (tx) => {
      const newUser = await tx.user.create({
        data: {
          email: input.email,
          passwordHash,
          firstName: input.firstName,
          lastName: input.lastName,
          role: input.role,
          phone: input.phone,
          profile: {
            create: {},
          },
        },
        select: {
          id: true,
          email: true,
          firstName: true,
          lastName: true,
          role: true,
          createdAt: true,
        },
      });

      // Create provider profile if role is therapist/psychiatrist/coach
      if (['THERAPIST', 'PSYCHIATRIST', 'COACH'].includes(input.role)) {
        await tx.providerProfile.create({
          data: {
            userId: newUser.id,
            providerType: input.role as any,
            licenseNumber: '', // To be filled during onboarding
            licenseBody: '',
            languages: [],
            sessionRate: 0,
          },
        });
      }

      // Create patient profile if role is patient
      if (input.role === 'PATIENT') {
        await tx.patientProfile.create({
          data: {
            userId: newUser.id,
          },
        });
      }

      return newUser;
    });

    // Generate tokens
    const { accessToken, refreshToken } = await this.generateTokenPair(user.id, user.email, user.role);

    logger.info('User registered successfully', { userId: user.id });

    return {
      user,
      accessToken,
      refreshToken,
    };
  }

  /**
   * Login user
   */
  async login(input: LoginInput) {
    // Find user
    const user = await prisma.user.findUnique({
      where: { email: input.email },
    });

    if (!user) {
      throw new AppError('Invalid credentials', HTTP_STATUS.UNAUTHORIZED, ERROR_CODES.INVALID_CREDENTIALS);
    }

    // Verify password
    const isValid = await cryptoUtils.verifyPassword(input.password, user.passwordHash);
    if (!isValid) {
      throw new AppError('Invalid credentials', HTTP_STATUS.UNAUTHORIZED, ERROR_CODES.INVALID_CREDENTIALS);
    }

    // Check if user is active
    if (!user.isActive) {
      throw new AppError('Account is deactivated', HTTP_STATUS.FORBIDDEN, ERROR_CODES.FORBIDDEN);
    }

    // Generate tokens
    const { accessToken, refreshToken } = await this.generateTokenPair(
      user.id,
      user.email,
      user.role,
      input.ipAddress,
      input.userAgent
    );

    // Update last login
    await prisma.user.update({
      where: { id: user.id },
      data: { lastLoginAt: new Date() },
    });

    logger.info('User logged in successfully', { userId: user.id });

    return {
      user: {
        id: user.id,
        email: user.email,
        firstName: user.firstName,
        lastName: user.lastName,
        role: user.role,
      },
      accessToken,
      refreshToken,
    };
  }

  /**
   * Refresh access token (with refresh token rotation)
   */
  async refreshToken(input: RefreshTokenInput) {
    // Verify refresh token
    let payload: TokenPayload;
    try {
      payload = jwt.verify(input.refreshToken, env.JWT_REFRESH_SECRET) as TokenPayload;
    } catch (error) {
      throw new AppError('Invalid refresh token', HTTP_STATUS.UNAUTHORIZED, ERROR_CODES.TOKEN_INVALID);
    }

    // Check if refresh token exists in DB
    const storedToken = await prisma.refreshToken.findUnique({
      where: { token: input.refreshToken },
      include: { user: true },
    });

    if (!storedToken) {
      throw new AppError('Refresh token not found', HTTP_STATUS.UNAUTHORIZED, ERROR_CODES.TOKEN_INVALID);
    }

    // Check if token is revoked
    if (storedToken.revokedAt) {
      throw new AppError('Refresh token has been revoked', HTTP_STATUS.UNAUTHORIZED, ERROR_CODES.TOKEN_INVALID);
    }

    // Check if token is expired
    if (storedToken.expiresAt < new Date()) {
      throw new AppError('Refresh token expired', HTTP_STATUS.UNAUTHORIZED, ERROR_CODES.TOKEN_EXPIRED);
    }

    // ROTATION: Revoke old refresh token and generate new pair
    await prisma.refreshToken.update({
      where: { id: storedToken.id },
      data: { revokedAt: new Date() },
    });

    // Generate new token pair
    const { accessToken, refreshToken: newRefreshToken } = await this.generateTokenPair(
      storedToken.user.id,
      storedToken.user.email,
      storedToken.user.role,
      input.ipAddress,
      input.userAgent
    );

    logger.info('Refresh token rotated', { userId: storedToken.user.id });

    return {
      accessToken,
      refreshToken: newRefreshToken,
    };
  }

  /**
   * Logout user
   */
  async logout(input: { userId: string; refreshToken?: string }) {
    if (input.refreshToken) {
      // Revoke specific refresh token
      await prisma.refreshToken.updateMany({
        where: {
          userId: input.userId,
          token: input.refreshToken,
        },
        data: {
          revokedAt: new Date(),
        },
      });
    } else {
      // Revoke all refresh tokens for user
      await prisma.refreshToken.updateMany({
        where: { userId: input.userId },
        data: { revokedAt: new Date() },
      });
    }

    logger.info('User logged out', { userId: input.userId });
  }

  /**
   * Get user by ID
   */
  async getUserById(userId: string) {
    const user = await prisma.user.findUnique({
      where: { id: userId },
      select: {
        id: true,
        email: true,
        firstName: true,
        lastName: true,
        role: true,
        avatarUrl: true,
        emailVerified: true,
        phoneVerified: true,
        isActive: true,
        createdAt: true,
      },
    });

    if (!user) {
      throw new AppError('User not found', HTTP_STATUS.NOT_FOUND, ERROR_CODES.RESOURCE_NOT_FOUND);
    }

    return user;
  }

  /**
   * Generate JWT token pair (access + refresh)
   */
  private async generateTokenPair(
    userId: string,
    email: string,
    role: UserRole,
    ipAddress?: string,
    userAgent?: string
  ) {
    const payload: TokenPayload = {
      userId,
      email,
      role,
    };

    // Generate access token
    const accessToken = jwt.sign(payload, env.JWT_ACCESS_SECRET, {
      expiresIn: env.JWT_ACCESS_EXPIRY,
    });

    // Generate refresh token
    const refreshToken = jwt.sign(payload, env.JWT_REFRESH_SECRET, {
      expiresIn: env.JWT_REFRESH_EXPIRY,
    });

    // Store refresh token in DB
    const expiresAt = new Date();
    expiresAt.setDate(expiresAt.getDate() + 7); // 7 days

    await prisma.refreshToken.create({
      data: {
        userId,
        token: refreshToken,
        expiresAt,
        ipAddress,
        userAgent,
      },
    });

    return { accessToken, refreshToken };
  }
}
